mlflow/types/schema.py

```diff
@@ -122,34 +125,280 @@ def from_numpy_type(cls, np_type):
         return next((v for v in cls._member_map_.values() if v.to_numpy() == np_type), None)
 
 
+ArrayType = TypeVar("ArrayType", bound="Array")
+ObjectType = TypeVar("ObjectType", bound="Object")
+
+
+@experimental
+class Property:
+    """
+    Specification used to represent a json-convertible object property.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        dtype: Union[DataType, ArrayType, ObjectType, str],  # pylint: disable=redefined-builtin
+        required: bool = True,
+    ) -> None:
+        """
+        :param name: The name of the property
+        :param dtype: The data type of the property
+        :param required: Whether this property is required
+        """
+        if not isinstance(name, str):
+            raise MlflowException.invalid_parameter_value(
+                f"Expected name to be a string, got type {type(name).__name__}"
+            )
+        self._name = name
+        try:
+            self._dtype = DataType[dtype] if isinstance(dtype, str) else dtype
+        except KeyError:
+            raise MlflowException(
+                f"Unsupported type '{dtype}', expected instance of DataType, Array, Object or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.dtype, (DataType, Array, Object)):
+            raise MlflowException(
+                "Expected mlflow.types.schema.Datatype, mlflow.types.schema.Array, "
+                "mlflow.types.schema.Object or str for the 'dtype' "
+                f"argument, but got {self.dtype.__class__}"
+            )
+        self._required = required
+
+    @property
+    def name(self) -> str:
+        """The property name."""
+        return self._name
+
+    @property
+    def dtype(self) -> Union[DataType, ArrayType, ObjectType]:
+        """The property data type."""
+        return self._dtype
+
+    @property
+    def required(self) -> bool:
+        """Whether this property is required"""
+        return self._required
+
+    @required.setter
+    def required(self, value: bool) -> None:
+        self._required = value
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Property):
+            return self.dtype == other.dtype and self.required == other.required
+        return False
+
+    def to_dict(self):
+        d = {"type": self.dtype.name} if isinstance(self.dtype, DataType) else self.dtype.to_dict()
+        d["required"] = self.required
+        return {self.name: d}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain only one key as `name`, and
+        the value should be a dictionary containing `type` and
+        optional `required` keys.
```

Comment:
It should look like this?

```
{
  "name": {"type": ..., "requierd": ...}
}
```

Reply:
can we add an example?

Reply:
Any reason that the following dict is not acceptable?

```
{
    "name": ...,
    "type": ...,
    "requierd": ...,
}
```

Reply:
To be consistent with the comment here: https://github.com/mlflow/mlflow/pull/9893#discussion_r1356049271
When we convert it to dictionary, we extract the keys out, so when we read from json, we should expect the same format right? Otherwise it gets quite confused that the to_dict and from_json_dict accepts different format.

Reply:
That makes sense, can we add an example in the docstring?